packages/core/fs/src/MemoryFS.js

```diff
@@ -0,0 +1,378 @@
+    dir = this._normalizePath(dir);
+    if (!this.dirs.has(dir)) {
+      throw new Error(`ENOENT: ${dir} does not exist`);
+    }
+
+    dir += path.sep;
+
+    let res = [];
+    for (let filePath of this.files.keys()) {
+      if (filePath.startsWith(dir)) {
+        let end = filePath.indexOf(path.sep, dir.length);
+        if (end === -1) {
+          end = filePath.length;
+        }
+        res.push(filePath.slice(dir.length, end));
+      }
+    }
+
+    return res;
+  }
+
+  async unlink(filePath: FilePath) {
+    filePath = this._normalizePath(filePath);
+    if (!this.files.has(filePath) && !this.dirs.has(filePath)) {
+      throw new Error(`${filePath} does not exist`);
+    }
+
+    this.files.delete(filePath);
+    this.dirs.delete(filePath);
+  }
+
+  async mkdirp(dir: FilePath) {
+    dir = this._normalizePath(dir);
+    if (this.dirs.has(dir)) {
+      return;
+    }
+
+    if (this.files.has(dir)) {
+      throw new Error('ENOTDIR: Not a directory');
+    }
+
+    let root = path.parse(dir).root;
+    while (dir !== root) {
+      if (this.dirs.has(dir)) {
+        break;
+      }
+
+      this.dirs.add(dir);
+      dir = path.dirname(dir);
+    }
+  }
+
+  async rimraf(filePath: FilePath) {
+    filePath = this._normalizePath(filePath);
+
+    if (this.dirs.has(filePath)) {
+      let dir = filePath + path.sep;
+      for (let filePath of this.files.keys()) {
+        if (filePath.startsWith(dir)) {
+          this.files.delete(filePath);
+        }
+      }
+
+      for (let dirPath of this.dirs.keys()) {
+        if (dirPath.startsWith(dir)) {
+          this.dirs.delete(dirPath);
+        }
+      }
+    }
+
+    this.dirs.delete(filePath);
+    this.files.delete(filePath);
+  }
+
+  async ncp(source: FilePath, destination: FilePath) {
+    source = this._normalizePath(source);
+
+    if (this.dirs.has(source)) {
+      this.dirs.add(destination);
+
+      let dir = source + path.sep;
+      for (let dirPath of this.dirs) {
+        if (dirPath.startsWith(dir)) {
+          let destName = path.join(destination, dirPath.slice(dir.length));
+          this.dirs.add(destName);
+        }
+      }
+
+      for (let [filePath, buffer] of this.files) {
+        if (filePath.startsWith(dir)) {
+          let destName = path.join(destination, filePath.slice(dir.length));
+          this.files.set(destName, buffer);
+        }
+      }
+    } else {
+      await this.copyFile(source, destination);
+    }
+  }
+
+  createReadStream(filePath: FilePath) {
+    let fs = this;
+    class ReadStream extends Readable {
+      _read() {
+        fs.readFile(filePath).then(
+          res => {
+            this.push(res);
+            this.push(null);
+          },
+          err => {
+            this.emit('error', err);
+          }
+        );
+      }
+    }
+
+    return new ReadStream();
+  }
+
+  createWriteStream(filePath: FilePath) {
+    let fs = this;
+    let buffer = Buffer.alloc(0);
+    class WriteStream extends Writable {
```

Comment:
- The `bytesWritten` property is missing:
https://github.com/parcel-bundler/parcel/blob/5a9b703a3f42c8fa3ea2df78100a85d5b6375237/packages/core/core/src/PackagerRunner.js#L181-L188
- It might be faster to not create a new class on every `createWriteStream` call?